### ISSUE TITLE
Request: launchers

### ISSUE THREAD

**Issue by luluco250** [lutris/lutris#4648]
Apologies upfront if this is something that's been proposed before or is a milestone target, I could not find anything similar to it.

**tl;dr:**
The idea is to be able to configure tools, programs or plain custom shortcuts for a game, besides the runner configuration itself.
The simplest use case is mod managers or similar tools, where you want to work on the exact same prefix the game uses.

Currently there are two ways I see to do it in Lutris:
1. Changing the executable/command line temporarily to the tool's, running the game, then going back and reverting it. This guarantees it uses the same configuration, but is pretty obviously annoying. Maybe a launcher could be seen as a way to automate this process.
1. For Wine runners, the option for running an EXE in the prefix. Convenient for one-time programs like installers and alike, but annoying for something you might to use often. A big miss from this is that certain configurations seem to be ignored, like environment variables.
1. Creating a separate entry for each tool you need. This can be more easily accomplished by duplicating the original entry and then changing what the runner executes, particularly useful for reusing the same Wine prefix. While it is a good solution, it quickly pollutes the game list with entries for game-specific tools, besides lacking the ability to automatically retrieve thumbnails.

I'd like to propose a third option, the ability to define custom "launchers" for games.
The way it could work is by having a list of programs and arguments you can add to a game's options, which would then appear in the context menu for that game (or the menu from the play button's arrow). These would then execute the command line for the launcher clicked, inheriting all or most of the game's Lutris configuration.

Examples:
- Bethesda games are notoriously moddable, so the use of a mod manager is very, very common. Although the more popular mod managers can launch the games directly, there are tools which won't (like LOOT or other more game-specific tools), so having the ability to quickly launch said tools through Lutris would be very useful and time-saving.
- Often you may want to skip launchers when playing games, but you might need to still use them from time to time to change a setting or two.
- Maybe a game has a different executable for multiplayer, or you use a mod that works over the game's original data but with a different executable. Maybe the game simply has different command line options for things like benchmarks.

There is a very clear cut analog to this in Steam, many games offer different executables/tools/modes when right clicked. Sometimes they are launchers, a special version of the game, maybe a multiplayer mode/server etc. I think that's a very good design to start from, except we'd need to figure out how to best approach letting users define launchers.

The way I see it, something similar to how game-specific settings inherit global settings could work well. Have each launcher start out copying the game's settings, then you can go ahead and change the executable, command line arguments, maybe disable things like gamescope or virtual desktop. But the idea is that we can minimize the amount of "reconfiguration" necessary, the intent is to run the launcher with something as close to the game as possible.

Like how a Windows installation may have many programs installed with quick and easy access, we could benefit from having this for Wine prefixes that may need different programs installed.

I understand this would probably be a pretty big task, so I'm just throwing the idea out there.
I could even start work on it myself if there isn't anything similar already being worked on, but I'd like to make sure I'm well aligned in relation to UI/UX, user expectations and the goals of the project before starting work on a PR.

**Comment by junglized** [comment]
There is a little game menu pop up, allowin to select game version, when launching Carmageddon: Max Damage (Steam or GOG) but I don't know how it's done. It's displayed as a native Linux window though so maybe this is coming from Lutris?

**Comment by danieljohnson2** [comment]
Probably you are seeing the launch configuration dialog. Installers (those .yml files) can set up alternate entry points for a game, and you get a dialog to select among them. There's no UI to set up your own interactively, but these launch configs do wind up in the game's configuration file.

Might be worth having such a UI; but certainly I'm not working on that. I do have a PR (#4582) around this feature but it's just a 'don't ask again' checkbox. Which is kind of the opposite of what you are thinking.

**Comment by junglized** [comment]
> Probably you are seeing the launch configuration dialog. Installers (those .yml files) can set up alternate entry points for a game, and you get a dialog to select among them. There's no UI to set up your own interactively, but these launch configs do wind up in the game's configuration file.
> 
> Might be worth having such a UI; but certainly I'm not working on that. I do have a PR (#4582) around this feature but it's just a 'don't ask again' checkbox. Which is kind of the opposite of what you are thinking.

But Carmageddon: Max Damage doesn't have installer script on Lutris' website. I installed it just by clicking an image in GOG source in Lutris and followed installation prompts.

**Comment by luluco250** [comment]
> Probably you are seeing the launch configuration dialog. Installers (those .yml files) can set up alternate entry points for a game, and you get a dialog to select among them. There's no UI to set up your own interactively, but these launch configs do wind up in the game's configuration file.
> 
> Might be worth having such a UI; but certainly I'm not working on that. I do have a PR (#4582) around this feature but it's just a 'don't ask again' checkbox. Which is kind of the opposite of what you are thinking.

I wasn't even aware of such a feature as I'm a bit of a control freak and don't use installers. It's good to know that there's at least some logic already implemented to support it, it just needs an UI to along with it.

But is that dialog window the only way to launch alternative entry points? Does the context/options menu also display them?

Personally I don't like how Steam will display a dialog when you try to launch a game as that disrupts the flow, specially when using controllers, so it'd be nice if it could also be a thing you deliberately have to go down a menu to find rather than blocking you and waiting for input.

I'll check out existing installers and maybe mess with the yaml directly to figure out how much of it suits my needs, thanks.

In any case, I might take on the task of implementing an UI for it, if the logic already exists in Lutris that cuts down on work a lot.

**Comment by danieljohnson2** [comment]
That dialog is the only way to use these entry points. The context menu is big enough; I don't think it should get this added into it.

I think a UI to edit these things would be very nice; I've just been struggling with incorrectly configured "launch_configs" from GOG, and with no UI to fix them it got a bit ugly.

**Comment by luluco250** [comment]
> That dialog is the only way to use these entry points. The context menu is big enough; I don't think it should get this added into it.
> 
> I think a UI to edit these things would be very nice; I've just been struggling with incorrectly configured "launch_configs" from GOG, and with no UI to fix them it got a bit ugly.

Well I didn't mean putting them on the first lebel of the context menu, but maybe like `Launchers > [ item1, item2 ... ]`. I don't think that would add significant clutter to it.

**Comment by CryoMyst** [comment]
This would be very useful to have.

**Comment by QmwJlHuSg9pa** [comment]
Tangentially related discussion (regarding Runner-level settings affecting main launch option vs. all launch options): https://github.com/lutris/lutris/issues/5148#issuecomment-1865928539

**Comment by drws** [comment]
Also related: #3566

**Comment by ItsAllAboutTheCode** [comment]
I have taken a stab at this feature and have a working protoptye.

I am looking for some feedback if this is the right approach.
Below is a video demonstrating how it works.

https://github.com/user-attachments/assets/8c2cb9d8-2aae-4c95-989a-904b788b8fe9

The  changes for the prototype are available in a branch at https://github.com/lutris/lutris/compare/master...ItsAllAboutTheCode:lutris:ui-for-launch-configs?expand=1

**Comment by SaltSouls** [comment]
> I have taken a stab at this feature and have a working protoptye.
> 
> I am looking for some feedback if this is the right approach. Below is a video demonstrating how it works.
> 
>  2026-03-25.19-52-01.mp4 
> The changes for the prototype are available in a branch at https://github.com/lutris/lutris/compare/master...ItsAllAboutTheCode:lutris:ui-for-launch-configs?expand=1

I was just having this issue with a game of mine, where it occasionally gets updates and needs to have the client updated; which is a separate executable. Having the ability to have selection built in natively would be awesome. Right now I had to create two separate games in Lutris, and then use a script to launch the desired game based on user input.
